app/org/sagebionetworks/bridge/dynamodb/DynamoNotificationRegistrationDao.java

```diff
@@ -0,0 +1,192 @@
+        // account there will be half a dozen, tops
+        return mapper.query(DynamoNotificationRegistration.class, query)
+                .stream().collect(Collectors.toList());
+    }
+
+    @Override
+    public NotificationRegistration getRegistration(String healthCode, String guid) {
+        checkNotNull(healthCode);
+        checkNotNull(guid);
+        
+        DynamoNotificationRegistration hashKey = new DynamoNotificationRegistration();
+        hashKey.setHealthCode(healthCode);
+        hashKey.setGuid(guid);
+        
+        NotificationRegistration registration = mapper.load(hashKey);
+        if (registration == null) {
+            throw new EntityNotFoundException(NotificationRegistration.class);
+        }
+        return registration;
+    }
+
+    @Override
+    public NotificationRegistration createRegistration(String platformARN, NotificationRegistration registration) {
+        checkNotNull(platformARN);
+        checkNotNull(registration);
+        checkNotNull(registration.getHealthCode());
+        checkNotNull(registration.getDeviceId());
+        checkNotNull(registration.getOsName());
+        
+        // (If the client is submitting same data a second time, SNS quietly ignores it, returns same endpointARN.) 
+        CreatePlatformEndpointRequest request = new CreatePlatformEndpointRequest()
+                .withToken(registration.getDeviceId())
+                .withCustomUserData(registration.getHealthCode())
+                .withPlatformApplicationArn(platformARN);
+        CreatePlatformEndpointResult result = snsClient.createPlatformEndpoint(request);
+        
+        // If the data is the same and returns an existing endpointARN, we want to re-use the original record 
+        // and GUID we provided to the client, not create a new record. Look for it.
+        NotificationRegistration existing = findExistingRecord(registration.getHealthCode(), result.getEndpointArn());
+        if (existing != null) {
+            registration.setGuid(existing.getGuid());
+            registration.setCreatedOn(existing.getCreatedOn());
+        } else {
+            registration.setGuid(BridgeUtils.generateGuid());    
+            registration.setCreatedOn(DateTime.now().getMillis());
+        }
+        registration.setHealthCode(registration.getHealthCode());
+        registration.setModifiedOn(DateTime.now().getMillis());
+        registration.setEndpointARN(result.getEndpointArn());
+        
+        mapper.save(registration);
+        return registration;
+    }
+    
+    /**
+     * Update an endpoint using the GUID that was supplied on creation. The only thing you can actually update 
+     * is the device token, but this is important to be able to update as it changes on some platforms.
+     */
+    @Override
+    public NotificationRegistration updateRegistration(String platformARN, NotificationRegistration registration) {
+        checkNotNull(platformARN);
```

Comment:
Okay, I'll remove. Makes sense that it's not used if we've already created this record.